File: architecture-skills/design-patterns/examples/facade/facade.go
// Facade pattern in Go.
//
// An idiomatic Go package *is* a facade: unexported types do the work
// and a small exported surface fronts them. Everything lives in one
// file here, so the "package boundary" is simulated with an exported
// HomeTheater over lower-case subsystem types.

package main

import "fmt"

type amplifier struct{}

func (amplifier) on()               { fmt.Println("Amplifier: on") }
func (amplifier) setVolume(lvl int) { fmt.Printf("Amplifier: volume %d\n", lvl) }
func (amplifier) off()              { fmt.Println("Amplifier: off") }

type projector struct{}

func (projector) on()             { fmt.Println("Projector: on") }
func (projector) wideScreenMode() { fmt.Println("Projector: 16:9 widescreen mode") }
func (projector) off()            { fmt.Println("Projector: off") }

type streamingPlayer struct{}

func (streamingPlayer) on() { fmt.Println("StreamingPlayer: on") }
func (streamingPlayer) play(movie string) {
	fmt.Printf("StreamingPlayer: playing %q\n", movie)
}
func (streamingPlayer) stop() { fmt.Println("StreamingPlayer: stopped") }
func (streamingPlayer) off()  { fmt.Println("StreamingPlayer: off") }

// HomeTheater is the facade: one method per use case.
type HomeTheater struct {
	amp    amplifier
	proj   projector
	player streamingPlayer
}

func NewHomeTheater() *HomeTheater {
	return &HomeTheater{}
}

func (h *HomeTheater) WatchMovie(movie string) {
	fmt.Println("= Movie night =")
	h.amp.on()
	h.amp.setVolume(5)
	h.proj.on()
	h.proj.wideScreenMode()
	h.player.on()
	h.player.play(movie)
}

func (h *HomeTheater) EndMovie() {
	fmt.Println("= Shutting down =")
	h.player.stop()
	h.player.off()
	h.proj.off()
	h.amp.off()
}

func main() {
	theater := NewHomeTheater()
	theater.WatchMovie("Inception")
	theater.EndMovie()
}

File: architecture-skills/design-patterns/examples/facade/facade.cpp
// Facade pattern in C++.
//
// The facade owns the subsystem objects by value and exposes one method
// per use case. No virtual functions are needed: clients depend on the
// facade concretely (the same idea underlies the pImpl idiom, which is
// a facade over a single translation unit).

#include <iostream>
#include <string>

class Amplifier {
public:
    void on() { std::cout << "Amplifier: on\n"; }
    void setVolume(int level) { std::cout << "Amplifier: volume " << level << '\n'; }
    void off() { std::cout << "Amplifier: off\n"; }
};

class Projector {
public:
    void on() { std::cout << "Projector: on\n"; }
    void wideScreenMode() { std::cout << "Projector: 16:9 widescreen mode\n"; }
    void off() { std::cout << "Projector: off\n"; }
};

class StreamingPlayer {
public:
    void on() { std::cout << "StreamingPlayer: on\n"; }
    void play(const std::string& movie) {
        std::cout << "StreamingPlayer: playing \"" << movie << "\"\n";
    }
    void stop() { std::cout << "StreamingPlayer: stopped\n"; }
    void off() { std::cout << "StreamingPlayer: off\n"; }
};

class HomeTheater {
public:
    void watchMovie(const std::string& movie) {
        std::cout << "= Movie night =\n";
        amp_.on();
        amp_.setVolume(5);
        projector_.on();
        projector_.wideScreenMode();
        player_.on();
        player_.play(movie);
    }

    void endMovie() {
        std::cout << "= Shutting down =\n";
        player_.stop();
        player_.off();
        projector_.off();
        amp_.off();
    }

private:
    Amplifier amp_;
    Projector projector_;
    StreamingPlayer player_;
};

int main() {
    HomeTheater theater;
    theater.watchMovie("Inception");
    theater.endMovie();
    return 0;
}
